### ISSUE TITLE
Simplify Golang definition once gopls is 1.0

### ISSUE THREAD

**Issue by lawrencegripper** [microsoft/vscode-dev-containers#375]
Golang has moved toward using `gopls` for nearly all language features. Most of the tooling installed in the current template is no longer needed and adds bloat to the container. 

Would it be alright to push another golang container and have two, one using the go language server and the other one using the existing setup?

With `gopls` the install drops to the following, with a few additional settings. I've also done in these examples is to pin the versions to avoid breaking changes, this feels like the right thing to do to preview breaking changes, for example #173 affecting users, but I understand this would incur more effort to maintain the template. 

```
RUN \
    # --> Delve for debugging
    go get github.com/go-delve/delve/cmd/dlv@v1.4.1 \
    # --> Go language server
    && go get golang.org/x/tools/gopls@v0.4.1 \
    # --> Go symbols and outline for go to symbol support and test support 
    && go get github.com/acroca/go-symbols@v0.1.1 && go get github.com/ramya-rao-a/go-outline@7182a932836a71948db4a81991a494751eccfe77
```
```
"settings": {
		"terminal.integrated.shell.linux": "/bin/bash",
		"go.gopath": "/go",
		"go.useLanguageServer": true,
		"[go]": {
			"editor.snippetSuggestions": "none",
			"editor.formatOnSave": true,
			"editor.codeActionsOnSave": {
				"source.organizeImports": true,
			}
		},
		"gopls": {
			"usePlaceholders": true, // add parameter placeholders when completing a function
			// Experimental settings
			"completeUnimported": true, // autocomplete unimported packages
			"watchFileChanges": true, // watch file changes outside of the editor
			"deepCompletion": true, // enable deep completion
		},
		"files.eol": "\n", // formatting only supports LF line endings		
	},
```

The repository container I pushed in this PR is an example of how much this can be slimmed down. https://github.com/microsoft/vscode-dev-containers/pull/374

Or similar in the [azbrowse devcontainer](https://github.com/lawrencegripper/azbrowse/blob/master/.devcontainer/Dockerfile).

**Comment by Chuxel** [comment]
@lawrencegripper I would prefer not to have two since customers would need to know which one to use in a given case. These show up in VS Code UX, so showing two with the same functionality won't make sense. With Google picking up management of the extension, I'd it also may make sense to have the extension owners weigh in on what is done here at this point. I'll see if I can get a contact.

**Comment by hyangah** [comment]
Is there any reason to hardcode the settings for gopls params? They look different from the gopls default settings.

`gopls` is still in alpha (so sometime users switch back and forth) and the `"gopls"` section in particular evolves very quickly. So, I'd prefer not to introduce an additional maintenance burden until `"gopls"` is close to 1.0. @stamblerre

And the default set by `gopls` when the section is not set must be good in general cases IMHO.

**Comment by Chuxel** [comment]
> Is there any reason to hardcode the settings for gopls params? They look different from the gopls default settings.

@hyangah Yeah, the settings in `devcontainer.json` only affect the extension when it is running in the container. (These are the settings under the "Remote" tab in the settings editor.) We've been adding them to the pre-defined definitions here since the paths could have been updated by the user to paths on their local machine - which in many cases will not work (e.g. Windows paths).  

We can set them to `null` to force the defaults to kick in, but the explicit paths make it a bit clearer as to what is going to happen.

**Comment by Chuxel** [comment]
@hyangah Forgot to say - for preferences, we generally do not set them in `devcontainer.json` unless it is a mode the container needs.  Those can flow through... it's just paths that typically are set. So, `"go.gopath": "/go"`

**Comment by hyangah** [comment]
@Chuxel thanks for the clarification. I like the idea of pinning the default versions of underlying go tools in general. (also, it allows us to use container caching in our ci :-) but I will let the gopls team and @stamblerre chime in since gopls is moving fast and vscode-go has a version check logic to always recommend the latest version.

**Comment by stamblerre** [comment]
I would prefer that we not pin versions of `gopls`, at least not until v1.0. Like @hyangah said, things are moving pretty fast right now, and it's important that we're able to get latest improvements to users quickly.

Similarly, the settings are frequently changing, and I think the only required setting to enable `gopls` should be `"go.useLanguageServer": true`. All of the settings listed above in the `gopls` section are now out-of-date, and `files.eol` is no longer needed either.

**Comment by Chuxel** [comment]
@hyangah @stamblerre Makes sense. Let's mark this as a feature request and tweak the description to specifically mention gopls going 1.0 to re-evaluate. That work?

**Comment by hyangah** [comment]
Yes, that works for us! Thanks!!

**Comment by lawrencegripper** [comment]
Sounds great - look forward to it, thanks all! 

Ps. Sorry for muddying the waters with my settings snipped, copied my current working one - agree many of these are out of date (woops) and personal preference.

@Chuxel on a related note, is there a way to pin an extension version in a devcontainer? We use them a fair bit on our team and saw the move to `golang.go` for the golang extension break all of those which didn't feel great - out aim was by pinning dependency versions our devcontainers wouldn't rot and we could pickup older projects easily, this broke that expectation a bit. For avoidance of doubt - It's the right move and I'm really happy that golang team are taking it on, would be nice option to say:

For example the `.devcontainer.json` could be something like this:
```
	// Add the IDs of extensions you want installed when the container is created in the array below.
	"extensions": [
		{ name: "golang.go", version: "~1.2"}
	]
```

**Comment by stuartleeks** [comment]
+1 One of the big benefits that I see for devcontainers is consistency of environments across a team. We hit issues a few times where this wasn't the case because one person built their devcontainer yesterday and then someone else built their devcontainer the following day, just after a release of one of the dependencies. When there are behavioural changes in the dependency this can cause a previously working environment to stop working. It wasn't immediately obvious what the issue was, but we have started pinning versions of dependencies in devcontainer definitions. (It was enough of an issue that it made it to [my list of thoughts on devcontainers ](https://stuartleeks.com/posts/vscode-devcontainers/#pin-your-tool-versions) :smile:)

**Comment by Chuxel** [comment]
> @Chuxel on a related note, is there a way to pin an extension version in a devcontainer? We use them a fair bit on our team and saw the move to golang.go for the golang extension break them and cause confusion - our aim was by pinning dependency versions our devcontainers wouldn't rot and we could pickup older projects easily, this broke that expectation a bit. For avoidance of doubt - It's the right move and I'm really happy that golang team are taking it on.

That is good feedback. VS Code generally follows a model where this is not desirable, but I can see in some cases that it would be. Would you mind raisning a feature request at https://github.com/microsoft/vscode-remote-release ?

> One of the big benefits that I see for devcontainers is consistency of environments across a team. 

@stuartleeks Yes, agreed. However, I think one of the challenges here (specifically) is maintenance since each tool can update independently and as mentioned above, `gopls` is still evolving. One of the reasons that these files get added to your project/repo instead of being referenced by ID is so you can opt into pinning if you find the need.

Perhaps one of the things that we could do is add build args so that the tool versions can be set from devcontainer.json and default them to latest to make updating them easier?

**Comment by lawrencegripper** [comment]
Sure, I'll raise the extension versioning as separate issue. 

> Perhaps one of the things that we could do is add build args so that the tool versions can be set from devcontainer.json and default them to latest to make updating them easier?

Yes I really like this idea so at the top of the `dockerfile` there would be a block of versions that can be set easily but default to latest. It also highlights that there is indeterminate behavior by default and provides a nice way to pin, if users want that, without having the maintenance burden on this repo.

**Comment by Chuxel** [comment]
I think this has been completed. We've removed anything the extension no longer needs.  LMK if you still think more is required, but closing as done.
